### PR TITLE
extras: add 'RutField' for Django forms

### PR DESCRIPTION
**Pull request by yaselc**
A separate field class for Django form

Previous: https://github.com/fyntex/lib-cl-sii-python/pull/5

### CODE REVIEW

**Comment by glarrain** [comment]
LGTM. Alas, the approval must come from @jtrh

File: tests/test_extras_dj_form_fields.py [hunk 1]
```diff
@@ -0,0 +1,65 @@
+import unittest
+
+import django.core.exceptions
+import django.forms  # noqa: F401
```
Comment (glarrain) [inline tests/test_extras_dj_form_fields.py:4]: not necessary

File: cl_sii/extras/dj_form_fields.py [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+"""
+try:
+    import django
+except ImportError as exc:  # pragma: no cover
+    raise ImportError("Package 'Django' is required to use this module.") from exc
+
+from typing import Any, Optional
+
+import django.core.exceptions
+import django.db.models
```
Comment (glarrain) [inline cl_sii/extras/dj_form_fields.py:13]: not necessary

File: cl_sii/extras/dj_form_fields.py [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+import django.db.models.fields
```
Comment (glarrain) [inline cl_sii/extras/dj_form_fields.py:14]: not necessary

File: cl_sii/extras/dj_form_fields.py [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+
+    * Python data type: :class:`cl_sii.rut.Rut`
+
+    .. seealso::
+        :class:`.dj_model_fields.RutField`
+
+    """
+
+    default_error_messages = {
+        'invalid': _('Enter a valid RUT.'),
```
Comment (glarrain) [inline cl_sii/extras/dj_form_fields.py:34]: I don't know whether we'll want to have support for translation or not, or if it's just a bonus. @yaselc @jtrh your call

File: cl_sii/extras/dj_form_fields.py [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+    }
+
+    def __init__(self, *, validated_dv: bool = False, **kwargs: Any) -> None:
```
Comment (glarrain) [inline cl_sii/extras/dj_form_fields.py:37]: why not an invalid DV message like in #5 ?

File: cl_sii/extras/dj_form_fields.py [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+        """
+        :param validated_dv: Boolean that specifies whether the RUT's
+            "digito verificador" should be validated. False by default.
+        """
+
+        self.validated_dv = validated_dv
```
Comment (glarrain) [inline cl_sii/extras/dj_form_fields.py:43]: Shouldn't `validated_dv` be named `validate_dv`?

File: tests/test_extras_dj_form_fields.py (hunk 1 above)
Comment (yaselc) [reply tests/test_extras_dj_form_fields.py]: Fixed

File: cl_sii/extras/dj_form_fields.py (hunk 2 above)
Comment (yaselc) [reply cl_sii/extras/dj_form_fields.py]: Fixed

File: cl_sii/extras/dj_form_fields.py (hunk 3 above)
Comment (yaselc) [reply cl_sii/extras/dj_form_fields.py]: Fixed

File: cl_sii/extras/dj_form_fields.py (hunk 4 above)
Comment (jtrh) [reply cl_sii/extras/dj_form_fields.py:32]: > I don't know whether we'll want to have support for translation or not, or if it's just a bonus.

I think that someday we should, at least for Spanish

Regarding the change itself, if `_('Enter a valid RUT.')` doesn't cause any additional work to our library users (e.g. by forcing them to execute `django-admin makemessages` and `django-admin compilemessages`), then I think it's fine.

File: cl_sii/extras/dj_form_fields.py (hunk 6 above)
Comment (jtrh) [reply cl_sii/extras/dj_form_fields.py:41]: I agree that it should be `validate_dv` (without the `d` after `validate`), unless there's some precedent to the contrary.

File: cl_sii/extras/dj_form_fields.py [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+        elif isinstance(value, Rut):
+            converted_value = value
+            if self.validated_dv and Rut.calc_dv(value.digits) != value.dv:
+                raise django.core.exceptions.ValidationError(
+                    self.error_messages['invalid'],
+                    code='invalid',
+                )
+        else:
+            try:
+                converted_value = Rut(value, validate_dv=self.validated_dv)  # type: ignore
```
Comment (jtrh) [inline cl_sii/extras/dj_form_fields.py:63]: Suggestion: Replace `# type: ignore` with `# type: ignore[arg-type]`.

> It’s possible to silence specific error codes on a line using # type: ignore[code]. This way you won’t accidentally ignore other, potentially more serious errors.

Source: [Error codes - Mypy Documentation](https://mypy.readthedocs.io/en/stable/error_codes.html#error-codes)

**Review by jtrh — APPROVED** [review]
There're a few suggestions in the review comments.

File: cl_sii/extras/dj_form_fields.py (hunk 5 above)
Comment (glarrain) [reply cl_sii/extras/dj_form_fields.py]: @yaselc ?

**Comment by glarrain** [comment]
@yaselc remember https://github.com/fyntex/lib-cl-sii-python/pull/197#discussion_r579212524

File: cl_sii/extras/dj_form_fields.py (hunk 5 above)
Comment (yaselc) [reply cl_sii/extras/dj_form_fields.py]: @glarrain do you mean using a separate error message for the case where the DV is not valid? e.g.

```python
default_error_messages = {
        'invalid': _('Enter a valid RUT.'),
        'invalid_dv': _('RUT\'s "digito verificador" is incorrect.')
    }
```

Comment (yaselc) [reply cl_sii/extras/dj_form_fields.py]: @glarrain? 
@jtrh can you help shed some light on this issue?

Comment (jtrh) [reply cl_sii/extras/dj_form_fields.py]: > @glarrain do you mean using a separate error message for the case where the DV is not valid?

I think that is what @glarrain meant.

You would need to change the `ValidationError` message in https://github.com/fyntex/lib-cl-sii-python/pull/197/files#diff-f6d12abbd698f974b0356078c32cbfcdb4b570857bf409850df17b2475b8cec0R56-R60 and differentiate between different error causes in https://github.com/fyntex/lib-cl-sii-python/pull/197/files#diff-f6d12abbd698f974b0356078c32cbfcdb4b570857bf409850df17b2475b8cec0R62-R68.

**Review by glarrain — APPROVED** [review]

**Review by jtrh — APPROVED** [review]

**Pull request merged**